cedar-lean/Cedar/Thm/Lemmas/Typechecker.lean

```diff
@@ -0,0 +1,175 @@
+definitions are an informed guess based on the corresponding Dafny proof.
+-/
+
+namespace Cedar.Thm
+
+open Cedar.Data
+open Cedar.Spec
+open Cedar.Validation
+
+def InstanceOfBoolType : Bool → BoolType → Prop
+  | true,  .tt      => True
+  | false, .ff      => True
+  | _,     .anyBool => True
+  | _, _            => False
+
+def InstanceOfEntityType (e : EntityUID) (ety: EntityType) : Prop :=
+  ety = e.ty
+
+def InstanceOfExtType : Ext → ExtType → Prop
+  | .decimal _, .decimal => True
+  | .ipaddr _,  .ipAddr  => True
+  | _, _                 => False
+
+inductive InstanceOfType : Value → CedarType → Prop :=
+  | instance_of_bool (b : Bool) (bty : BoolType)
+      (h₁ : InstanceOfBoolType b bty) :
+      InstanceOfType (.prim (.bool b)) (.bool bty)
+  | instance_of_int :
+      InstanceOfType (.prim (.int _)) .int
+  | instance_of_string :
+      InstanceOfType (.prim (.string _)) .string
+  | instance_of_entity (e : EntityUID) (ety: EntityType)
+      (h₁ : InstanceOfEntityType e ety) :
+      InstanceOfType (.prim (.entityUID e)) (.entity ety)
+  | instance_of_set (s : Set Value) (ty : CedarType)
+      (h₁ : forall v, v ∈ s → InstanceOfType v ty) :
+      InstanceOfType (.set s) (.set ty)
+  | instance_of_record (r : Map Attr Value) (rty : RecordType)
+      -- if an attribute is present, then it has the expected type
+      (h₁ : ∀ (k : Attr) (v : Value) (qty : QualifiedType),
+        rty.find? k = some qty → r.find? k = some v → InstanceOfType v qty.getType)
+      -- required attributes are present
+      (h₂ : ∀ (k : Attr) (qty : QualifiedType), rty.find? k = some qty → qty.isRequired → r.contains k) :
+      InstanceOfType (.record r) (.record rty)
+  | instance_of_ext (x : Ext) (xty : ExtType)
+      (h₁ : InstanceOfExtType x xty) :
+      InstanceOfType (.ext x) (.ext xty)
+
+def InstanceOfRequestType (request : Request) (reqty : RequestType) : Prop :=
+  InstanceOfEntityType request.principal reqty.principal ∧
+  request.action = reqty.action ∧
+  InstanceOfEntityType request.resource reqty.resource ∧
+  InstanceOfType request.context (.record reqty.context)
+
+/--
+For every entity in the store,
+1. The entity's type is defined in the type store.
+2. The entity's attributes match the attribute types indicated in the type store.
+3. The entity's ancestors' types are consistent with the ancestor information
+   in the type store.
+-/
+def InstanceOfEntityTypeStore (entities : Entities) (ets: EntityTypeStore) : Prop :=
+  ∀ uid data, entities.find? uid = some data →
+    ∃ attrTys ancestorTys, ets.find? uid.ty = some (attrTys, ancestorTys) ∧
+      InstanceOfType data.attrs (.record attrTys) ∧
+      ∀ ancestor, ancestor ∈ data.ancestors → ancestor.ty ∈ ancestorTys
+
+/--
+For every action in the entity store, the action's ancestors are consistent
+with the ancestor information in the action store.
+-/
+def InstanceOfActionStore (entities : Entities) (as: ActionStore) : Prop :=
+  ∀ uid data, entities.find? uid = some data →
```

Comment:
Does this need something to limit `uid` to those that are actions?

Reply:
Good catch.  I think it needs to be something along the lines of 

```
∀ uid data, entities.find? uid = some data → as.contains uid → ...
```

Reply:
Oops! Fixed.